[Cedric Ong] iP

cedricongjh opened the pull request:
# Duke

> This is a bot that helps you with tasks - Cedric Ong (Me)

It can keep track of:

* Todos
* Events
* Deadlines

All you need to do is:
1. Download the JAR file from this [repository](https://github.com/cedricongjh/ip)
2. Add your tasks
3.  Done!

Upcoming:
- [x] JUnit testing
- [ ] Gradle 
- [ ] Final release :tada:

---

For Java programmers that want to improve the code ~~please don't~~ here's the `main` method:

```java
public static void main(String[] args) {

    Parser parser = new Parser();
    UI ui = new UI(new Scanner(System.in));
    Storage storage = new Storage("./data/tasks.txt");
    TaskList tasks;

    try {
        tasks = new TaskList(storage.loadTasks());
    } catch (DukeException e) {
        ui.showException(e);
        tasks = new TaskList(new ArrayList<>());
    }

    ui.showWelcome();

    boolean isExit = false;

    while (!isExit) {
        try {
            String userInput = ui.getUserInput();
            Command c = parser.parse(userInput);
            c.execute(tasks, ui, storage);
            isExit = c.isExit();
        } catch (DukeException e) {
            ui.showException(e);
        }
    }
}
```

Commit:
Add Gradle support

cedricongjh committed:
add Greet, echo, exit functionality

cedricongjh committed:
added task list

cedricongjh committed:
added Task class

cedricongjh committed:
changed to switch statement and use Task class to mark as done

cedricongjh committed:
create subclasses for Task

cedricongjh committed:
add todos events and deadlines

cedricongjh committed:
add automated test

cedricongjh committed:
edit event to have at instead of by

cedricongjh committed:
add DukeException class

cedricongjh committed:
add basic exception handling to Duke

cedricongjh committed:
update test for exception handling

cedricongjh committed:
create method to get user inputted number

cedricongjh committed:
switch taskList from array to arraylist

cedricongjh committed:
remove task attribute from Duke

cedricongjh committed:
add deletion of task to Duke

cedricongjh committed:
add tests for task deletion

cedricongjh committed:
Add methods to format Task for saving to text file

cedricongjh committed:
Add methods for loading and saving tasks to file

cedricongjh committed:
Add data folder, and update gitignore to ignore tasks.txt

cedricongjh committed:
Add LocalDate to event and deadline

cedricongjh committed:
Merge pull request #1 from cedricongjh/branch-Level-7

Branch level 7

cedricongjh committed:
Merge branch 'master' of github.com:cedricongjh/ip into branch-Level-8

cedricongjh committed:
update EXPECTED.TXT to new date format

cedricongjh committed:
Merge pull request #2 from cedricongjh/branch-Level-8

Add LocalDate to event and deadline

cedricongjh committed:
Fix event toString method

cedricongjh committed:
Simplify constructors for Task, Todo, Event and Deadline

cedricongjh committed:
Fix dates not saving to text file

cedricongjh committed:
Create storage class

cedricongjh committed:
Create TaskList class

cedricongjh committed:
Create Parser and UI class

cedricongjh committed:
Organise classes into packages

cedricongjh committed:
Update path for Duke in runtest.bat

cedricongjh committed:
Initialise jUnit tests

cedricongjh committed:
Add javadocs to some public methods

cedricongjh committed:
Add MANIFEST.MF

cedricongjh committed:
Reformat code to comply with coding standard

cedricongjh committed:
Add find functionality

cedricongjh committed:
Merge pull request #3 from cedricongjh/A-CodingStandard

Reformat code to comply with coding standard

cedricongjh committed:
Merge branch 'master' into branch-Level-9

cedricongjh committed:
Merge pull request #4 from cedricongjh/branch-Level-9

Add find functionality

cedricongjh committed:
Merge pull request #5 from cedricongjh/branch-A-JavaDoc

Add javadocs to some public methods

cedricongjh committed:
Create Utils class

cedricongjh committed:
Create Command classes

cedricongjh committed:
Refactor Duke and Parser to make use of Command class

ZhaoPeiduo commented on src/main/java/duke/Utils.java:
```diff
@@ -0,0 +1,48 @@
+        StringJoiner end = new StringJoiner(" ");
+
+        int i = 0;
+        boolean after = false;
```
Maybe the boolean name should be changed to isAfter to meet the coding standard.

ZhaoPeiduo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,117 @@
+                if (index != taskIndex) {
+                    newTasks.append(fileReader.nextLine()).append("\n");
+                } else if (!delete) {
+                    newTasks.append(tasks.getTask(taskIndex).toFileData()).append("\n");
```
Maybe the boolean name should be changed to isDelete to meet the coding standard.

ZhaoPeiduo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,61 @@
+        String userCommand = getUserCommand(userInput);
+        String userArgument = getUserArgument(userInput);
+
+        switch (userCommand) {
```
Very clean switch block for the parse method!

ZhaoPeiduo commented on src/main/java/duke/command/TaskCommand.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public void execute(TaskList tasks, UI ui, Storage storage) throws DukeException {
+        if (userArgument.equals("")) {
+            throw new DukeException("The description of a duke.task.Task cannot be empty.");
+        }
```
I like the check for edge case that avoids a deep nested loop!

ZhaoPeiduo commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public void execute(TaskList tasks, UI ui, Storage storage) throws DukeException {
+        int delete = Utils.getInputNumber(userArgument);
```
Maybe it is better for the variable name to be indexToDelete?

ZhaoPeiduo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,117 @@
+    private static Task dataToTask(String str) throws DukeException {
+        String[] taskArr = str.split(",");
+        String taskType = taskArr[0];
+        boolean taskDone = taskArr[1].equals("1");
```
Maybe the boolean name should be changed to isTaskDone to meet the coding standard.

cedricongjh replied:
ok, thanks for pointing this out! as well as the other comments given

ZhaoPeiduo reviewed (commented):
As far as I can see, there is no violations for the code quality guideline(basic and intermediate levels). Maybe it is better to pay more attention to boolean variable naming as suggested by the comments!

cedricongjh committed:
Merge pull request #6 from cedricongjh/add-gradle-support

Add Gradle support

cedricongjh committed:
Specify correct mainClassName in build.gradle

muhammad-khair commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,25 @@
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Note that static imports are to be ordered above normal imports, contrary to the default settings in IntelliJ.  
See: https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements  

![image](https://user-images.githubusercontent.com/66059643/131636462-bd0e6583-a1b5-426d-97b1-cf45fe6aeda8.png)

muhammad-khair commented on src/test/java/duke/task/TaskTest.java:
```diff
@@ -0,0 +1,14 @@
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Similar to the previous comment on the ordering of imports.

muhammad-khair commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,16 @@
+
+public class Todo extends Task {
+
+    public Todo(String description, boolean done) { super(description, done); }
```
I am not so sure about having a single-lined method declaration and code, might need to check with others/prof/TAs about this.  
Referring to the Google Java Style Guide (for stuff not on the module site), we are recommended to break it up to:
``` java
public Todo(String description, boolean done) {
    super(description, done);
}
```
See: https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style

![image](https://user-images.githubusercontent.com/66059643/131637726-dd888b41-d8eb-4cc2-a680-91f2867e8afc.png)

muhammad-khair commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+        this.isDone = done;
+    }
+
+    public Task(String description, String date, boolean done) throws DukeException {
```
If I am not wrong, we still need to include Javadocs for public constructors. We can only ignore Javadocs for public getters and setters methods.
See: https://se-education.org/guides/conventions/java/intermediate.html#comments

muhammad-khair commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,29 @@
+package duke.command;
+
+import duke.DukeException;
```
If there are unused imports in the class, please remove them.

muhammad-khair commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,34 @@
+        doneTask.markAsDone();
+        storage.editTaskFromFile(done, tasks);
+
+        ui.showMessage(String.format("I've marked this task as done:\n" +
```
The `+` should be brought towards the next line, i.e.  
``` java
ui.showMessage(String.format("I've marked this task as done:\n"
        + "%s\n", doneTask.toString()));
``` 
See: https://se-education.org/guides/conventions/java/intermediate.html#layout

muhammad-khair commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,117 @@
+            taskDate = taskArr[3];
+        }
+        Task res = null;
+        switch (taskType) {
```
Note that `switch-case` indentation need to be different standard from default IntelliJ settings, i.e. it should be like this (similar to the indentation style you used in `Parser.java`):
``` java
switch (taskType) {
case("T"):
    res = new Todo(taskDescription, taskDone);
    break;
case("D"):
    res = new Deadline(taskDescription, taskDate, taskDone);
    break;
case("E"):
    res = new Event(taskDescription, taskDate, taskDone);
    break;
}
```  
See: https://se-education.org/guides/conventions/java/intermediate.html#layout > `switch statements`

![image](https://user-images.githubusercontent.com/66059643/131639297-8e704e0d-b0e6-44ce-8b1c-6a10fa14ca3f.png)

muhammad-khair commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,61 @@
+package duke;
+
+import duke.command.*;
```
Wild card imports `import duke.command.*` are not allowed in coding standards. To remove this auto-setting in IntelliJ, go to `Settings` > `Editor` > `Code Style` > `Java`, and set a super high number for `Class count to use import with '*'`. See bottom of the screenshot.  

![image](https://user-images.githubusercontent.com/66059643/131639728-1af1690e-acc7-476f-9e18-64052807a8a4.png)

muhammad-khair commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,117 @@
+            case("E"):
+                res = new Event(taskDescription, taskDate, taskDone);
+                break;
+        }
```
Just my opinion, but maybe a `default` block in `switch-case` to catch unknown edge cases would be nice if possible.

muhammad-khair reviewed (commented):
Overall, I find the code is relatively neat and there is a nice flow in logic during my browsing. However, there are some code standards issues that might need to be fixed. One way to aid you in making checks is using a `checkstyle.xml` file and `Gradle`.

To do this, you can go to https://github.com/se-edu/addressbook-level3/tree/master/config/checkstyle and obtain the `.XML` files needed, since it follows very similarly to our module requirements.  

Create the same directory path in your own `iP` repo (`repo/config/checkstyle/*.xml`). Then you can just use the `Gradle` panel > `Tasks` > `other` > `checkstyleMain` or `checkstyleTest` which will run and give you a report of code style violations (if any) in the terminal output (which gives a hyperlink to a local `HTML` page with the results).

All the best for the rest of your `iP` and **CS2103**!

cedricongjh committed:
Add FindCommand to Parser

cedricongjh committed:
Add dependancies for javaFx

cedricongjh committed:
Refactor Duke to remove text-based UI, and return a String response to any input instead

cedricongjh committed:
Delete UI class

cedricongjh committed:
Create basic GUI for Duke

cedricongjh committed:
Specify new Main class in build.gradle

cedricongjh committed:
Configure build.gradle to enable assertions when running Duke

By default, Java disables assertions. This might lead to one assuming
that the assertions are True, when they might not be.

Let's configure build.gradle to enable assertions such that any assertions
added in the future would run, and give the desired effect of using assertions.

cedricongjh committed:
Add assertions in certain commands to ensure userArgument is not null

It is possible for a developer to pass in null to the constructor of a Command
This is undesired as certain commands require that there be an either an empty
string, or a string as the argument.

Let's use assertions to ensure that future developers do not pass in null
as the userArgument to the Commands.

cedricongjh committed:
Add assertions in Duke to ensure filepath is not null

It is possible for a developer to pass in null to the constructor of Duke.
This is undesired as Duke requires a String as the filepath.

Let's use assertions to ensure that future developers do not pass in null
as the filepath to Duke.

cedricongjh committed:
Add assertions in Utils to ensure certain parameters are not null

It is possible for a developer to pass in null to the methods,
getInputNumber and showTasks. This is undesirable as it would result
in a NullPointerException.

Let's use assertions to ensure that future developers do not pass in null
as the filepath to Duke.

cedricongjh commented:
Hi, @muhammad-khair, just saw your comments today, really appreciate them! Will work to rectify the issues that you have brought up.

cedricongjh committed:
Remove duplicated splitBetween method in TaskCommand

cedricongjh committed:
Rename boolean in splitBetween to match coding standard

cedricongjh committed:
Change boolean variables and switch statement to follow coding style

cedricongjh committed:
Ensure dataToTask switch statement has a default that throws an error

cedricongjh committed:
Rename delete variable to indexToDelete

cedricongjh committed:
Change imports to meet coding standard

cedricongjh committed:
Reorder imports to meet coding standard

cedricongjh committed:
Format multiline string to meet coding standard

cedricongjh committed:
Define constants for commands, and replace strings with constants

cedricongjh committed:
Change ordering of final and static in constants

cedricongjh committed:
Create constants to remove magic literals for functions related to Tasks

cedricongjh committed:
Merge pull request #7 from cedricongjh/branch-A-CodeQuality

Code quality

cedricongjh committed:
Merge branch 'master' into branch-A-Assertions

cedricongjh committed:
Merge pull request #8 from cedricongjh/branch-A-Assertions

Assertions

cedricongjh committed:
Add getter for LocalDate in Task

cedricongjh committed:
Add method to sort TaskList

cedricongjh committed:
Add sorting to list command to show user a sorted list

cedricongjh committed:
Merge pull request #9 from cedricongjh/branch-C-Sort

Branch c sort

cedricongjh committed:
Prevent truncation of long responses

cedricongjh committed:
Make bye command close Duke window

cedricongjh committed:
Delete data file if the current file's data cannot be loaded

cedricongjh committed:
Refactor exceptions into its own package and create specific exceptions

cedricongjh committed:
Add CSS styling to improve GUI

cedricongjh committed:
Add warnings to user in case of Storage failing to save to file

cedricongjh committed:
Update avatar pictures

cedricongjh committed:
Refactor GUI components into its own package

cedricongjh committed:
Refactor Storage, Utils, Parser, TaskList into its own package

cedricongjh committed:
Add JavaDocs for Utils package, minor refactoring of code

cedricongjh committed:
Add JavaDocs for gui package, minor refactoring of code

cedricongjh committed:
Add JavaDocs for Duke, minor refactoring of code

cedricongjh committed:
Add JavaDocs for task package, minor refactoring of code

cedricongjh committed:
Add JavaDocs for exceptions package

cedricongjh committed:
Add JavaDocs for command package, minor code refactoring for invalid command

cedricongjh committed:
Fix code style issue

cedricongjh committed:
Finalise User and Duke profile picture

cedricongjh committed:
Add product screenshot

cedricongjh committed:
Update README.md

cedricongjh committed:
Update README.md
